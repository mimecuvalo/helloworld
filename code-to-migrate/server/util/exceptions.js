export class HTTPError extends Error {
  constructor(status, url, opt_info) {
    super(status.toString());

    opt_info = opt_info || '';
    this.name = `${status}: ${url} ${opt_info}`;
  }
}
