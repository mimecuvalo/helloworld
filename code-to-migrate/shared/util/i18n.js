import {
  FormattedMessage,
  createIntlCache,
  createIntl as originalCreateIntl,
  defineMessages as originalDefineMessages,
} from 'react-intl';

import MD5 from 'md5.js';

// Re-export everything and override below what we want to override.
export * from 'react-intl';

const SETTINGS = {
  defaultLocale: 'en',
  locales: ['en', 'fr'],
};

const INTERNAL_LOCALES = ['xx-AE', 'xx-LS'];

// This matches the extraction tool pattern:
//   --id-interpolation-pattern '[md5:contenthash:hex:10]'
function generateId({ id, description, defaultMessage }) {
  if (id) {
    return id;
  }

  return new MD5()
    .update(description ? `${defaultMessage}#${description}` : defaultMessage)
    .digest('hex')
    .slice(0, 10);
}

export function F(props) {
  const id = generateId(props);
  return (
    <span className="i18n-msg">
      <FormattedMessage id={id} {...props} />
    </span>
  );
}

// We programmatically define ID's for messages to make things easier for devs.
export function defineMessages(values) {
  for (const key in values) {
    if (!values[key].id) {
      values[key].id = generateId(values[key]);
    }
  }
  return originalDefineMessages(values);
}

export function setLocales({ defaultLocale, locales }) {
  SETTINGS.defaultLocale = defaultLocale;
  SETTINGS.locales = locales;
}

export function getDefaultLocale() {
  return SETTINGS.defaultLocale;
}

export function getLocales() {
  return SETTINGS.locales;
}

// Based on the request object (and other metrics, if you so chose) deduce the locale
// that the app should be rendered with.
export function getLocaleFromRequest(req) {
  // You can add logic here to extract a locale from your user object.
  // if (user.preferences.locale) {
  //   return findRelevantLocale(user.preferences.locale);
  // }

  // Language override via URL.
  if (req.query.lang) {
    const locale = findRelevantLocale(req.query.lang);
    if (locale) {
      return locale;
    }
  }

  // If not in user's preferences, we try to extract from the browser 'Accept-Language' header.
  if (req.headers['accept-language']) {
    const rawHeader = req.headers['accept-language'];
    const possibleLanguages = rawHeader.split(',').map((lang) => lang.replace(/;q=.*/, ''));
    for (const language of possibleLanguages) {
      const locale = findRelevantLocale(language);
      if (locale) {
        return locale;
      }
    }
  }

  // Final fallback
  return SETTINGS.defaultLocale;
}

// Find the exact match locale, if supported, or the next best locale if possible.
// e.g. if `fr-FR` isn't found then `fr` will be used.
function findRelevantLocale(locale) {
  if (isValidLocale(locale)) {
    return locale;
  }

  const baseLocale = locale.split('-')[0];
  if (isValidLocale(baseLocale)) {
    return baseLocale;
  }
}

// Whether the locale is found in our supported locale list. Must be exact.
export function isValidLocale(locale) {
  return SETTINGS.locales.indexOf(locale) !== -1 || isInternalLocale(locale);
}

export function isInternalLocale(locale) {
  return process.env.NODE_ENV === 'development' && INTERNAL_LOCALES.indexOf(locale) !== -1;
}

// This is optional but highly recommended since it prevents memory leaks.
// See: https://formatjs.io/docs/intl/#createintl
const cache = createIntlCache();
let presetIntl = null;
let didSetupCreateIntl = false;
export function setupCreateIntl({ defaultLocale, locale, messages }) {
  presetIntl = originalCreateIntl(
    {
      defaultLocale,
      locale,
      messages,
    },
    cache
  );

  didSetupCreateIntl = true;
}

export function createIntl(options) {
  if (options) {
    return originalCreateIntl(options);
  } else {
    if (!didSetupCreateIntl) {
      throw new Error('Need to run setupCreateIntl to use createIntl without options.');
    }
    return presetIntl;
  }
}
