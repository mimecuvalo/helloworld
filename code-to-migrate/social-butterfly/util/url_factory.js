export function buildUrl({ req, isAbsolute, pathname, searchParams }) {
  let url = '';

  if (req) {
    const protocol = req.get('x-scheme') || req.protocol;
    url += `${protocol}://${req.get('host')}`;
  } else if (isAbsolute) {
    url += `${window.location.origin}`;
  }

  url += !pathname || pathname.startsWith('/') ? pathname : new URL(pathname).pathname;

  if (searchParams) {
    url += '?' + new URLSearchParams(searchParams).toString();
  }

  return prettifyUrl(url);
}

export function prettifyUrl(url) {
  return url.replace(/ /g, '+');
}

export function ensureAbsoluteUrl(basisAbsoluteUrl, urlOrPath) {
  const parsedUrl = new URL(basisAbsoluteUrl);
  const hostnameAndProtocol = `${parsedUrl.protocol}//${parsedUrl.host}`;

  urlOrPath = urlOrPath || '';
  return urlOrPath[0] === '/' ? `${hostnameAndProtocol}${urlOrPath}` : urlOrPath;
}
