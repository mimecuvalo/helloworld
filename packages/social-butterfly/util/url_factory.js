export function buildUrl({ req, isAbsolute, pathname, searchParams }) {
  let url = '';

  if (req) {
    const protocol = req.get('x-scheme') || req.protocol;
    url += `${protocol}://${req.get('host')}`;
  } else if (isAbsolute) {
    url += `${window.location.origin}`;
  }

  url += !pathname || pathname.startsWith('/') ? pathname : new URL(pathname).pathname;

  if (searchParams) {
    url += '?' + new URLSearchParams(searchParams).toString();
  }

  return prettifyUrl(url);
}

export function prettifyUrl(url) {
  return url.replace(/ /g, '+');
}
