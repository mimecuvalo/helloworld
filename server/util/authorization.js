// This file tries to create a centralized place for authorization so that it can be used
// in the app, API, and Apollo servers.
export const isAuthenticated = (currentUser) => !!currentUser;
export const isAuthor = (currentUser) => !!currentUser?.model;
export const isAdmin = (currentUser) => !!currentUser?.model?.superuser;

const authorization = {
  isAuthenticated,
  isAuthor,
  isAdmin,
};
export default authorization;
