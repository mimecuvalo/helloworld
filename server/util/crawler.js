import fetch from 'node-fetch';
import { NotFoundError } from '../util/exceptions';

export async function fetchUrl(url) {
  try {
    const response = await fetch(url, {
      headers: {
        'user-agent': 'hello, world bot.',
      },
    });
    if (response.status !== 200) {
      throw new NotFoundError();
    }
    return response;
  } catch (ex) {
    throw ex;
  }
}

export function createAbsoluteUrl(websiteUrl, url) {
  if (url.startsWith('/')) {
    const parsedUrl = new URL(websiteUrl);
    url = `${parsedUrl.origin}${url}`;
  }

  return url;
}

export function isRobotViewing(req) {
  const userAgent = req.headers['x-user-agent'] || req.headers['user-agent'];
  return !!userAgent?.match(/bot|spider|crawl|slurp|ia_archiver/i);
}
