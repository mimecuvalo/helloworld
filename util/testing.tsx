import '@testing-library/jest-dom';
import { InMemoryCache, gql } from '@apollo/client';
import React, { FC, PropsWithChildren, ReactElement } from 'react';
import { RenderOptions, render } from '@testing-library/react';

import { IntlProvider } from 'i18n';
import { MockedProvider } from '@apollo/client/testing';
import { ThemeProvider } from '@mui/material/styles';
import { muiTheme } from 'styles';

const cache = new InMemoryCache();
cache.writeQuery({
  query: gql`
    query helloAndEchoQueries($str: String!) {
      echoExample(str: $str) {
        exampleField
      }

      hello
    }
  `,
  data: {
    echoExample: {
      __typename: 'Echo',
      exampleField: '/',
    },
    hello: 'GraphQL',
  },
});

const AllTheProviders: FC<PropsWithChildren> = ({ children }) => {
  //const apolloClient = useApollo(pageProps);

  return (
    <ThemeProvider theme={muiTheme}>
      <MockedProvider cache={cache}>
        {/* @ts-ignore looks like IntlProvider still needs updated types after React 18 transition. */}
        <IntlProvider defaultLocale="en" locale="en" messages={{}}>
          {/* @ts-ignore looks like IntlProvider still needs updated types after React 18 transition. */}
          {children}
        </IntlProvider>
      </MockedProvider>
    </ThemeProvider>
  );
};

const customRender = (ui: ReactElement<any>, options?: Omit<RenderOptions, 'wrapper'>) =>
  render(ui, { wrapper: AllTheProviders, ...options });

export * from '@testing-library/react';
export { customRender as render };
