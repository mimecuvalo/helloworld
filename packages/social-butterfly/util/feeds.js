import cheerio from 'cheerio';
import FeedParser from 'feedparser';
import { fetchText, fetchUrl, createAbsoluteUrl, sanitizeHTML } from './crawler';
import { HTTPError } from './exceptions';
import { Readable } from 'stream';

export async function discoverAndParseFeedFromUrl(url) {
  const { content, feedUrl } = await discoverAndRetrieveFeedFromUrl(url);
  const { feedEntries, feedMeta } = await parseFeed(content);

  return { feedEntries, feedMeta, feedUrl };
}

async function discoverAndRetrieveFeedFromUrl(url) {
  const response = await fetchUrl(url);
  const content = await response.text();

  const contentType = response.headers.get('content-type');
  if (contentType?.includes('text/html')) {
    return await parseHtmlAndRetrieveFeed(url, content);
  }

  // The url is the feed already, just send that back.
  return { content, feedUrl: url };
}

async function parseHtmlAndRetrieveFeed(websiteUrl, html) {
  const $ = cheerio.load(html);
  const links = $('link[rel="alternate"]').filter((index, el) => ($(el).attr('type') || '').match(/(rss|atom)/));

  let feedUrl = links.first().attr('href');
  if (!feedUrl) {
    throw new HTTPError(404, websiteUrl, 'feed: no feed url');
  }

  feedUrl = createAbsoluteUrl(websiteUrl, feedUrl);
  const content = await retrieveFeed(feedUrl);
  return { content, feedUrl };
}

export async function parseFeedAndInsertIntoDb(options, userRemote, feedResponseText, logger) {
  try {
    const { feedEntries } = await parseFeed(feedResponseText);
    await mapFeedAndInsertIntoDb(options, userRemote, feedEntries, logger);
  } catch (ex) {
    logger && logger.error(`${userRemote.local_username} - ${userRemote.profile_url}: parseFeed FAILED.\n${ex}`);
  }
}

export async function mapFeedAndInsertIntoDb(options, userRemote, feedEntries, logger) {
  let newEntries, skippedCount;
  try {
    [newEntries, skippedCount] = await mapFeedEntriesToModelEntries(options, feedEntries, userRemote);
    logger &&
      logger.info(
        `${userRemote.local_username} - ${userRemote.profile_url}: ` +
          `parsed ${newEntries.length} entries, skipped ${skippedCount}.`
      );
  } catch (ex) {
    logger && logger.error(`${userRemote.local_username} - ${userRemote.profile_url}: mapFeed FAILED.\n${ex}`);
    return;
  }

  try {
    newEntries.length && (await options.saveRemoteContent(newEntries));
    logger &&
      logger.info(
        `${userRemote.local_username} - ${userRemote.profile_url}: inserted ${newEntries.length} entries into db.`
      );
  } catch (ex) {
    logger &&
      logger.error(`${userRemote.local_username} - ${userRemote.profile_url}: db insertion failed.\n${ex.stack}`);
  }
}

export async function retrieveFeed(feedUrl) {
  return await fetchText(feedUrl);
}

export async function parseFeed(content) {
  const { feedEntries, feedMeta } = await new Promise((resolve, reject) => {
    const feedEntries = [];
    new TextStream({}, content)
      .pipe(new FeedParser())
      .on('error', function (error) {
        reject(`FeedParser failed to parse feed: ${error}`);
      })
      .on('readable', function () {
        try {
          let feedEntry = this.read();
          while (feedEntry) {
            feedEntries.push(feedEntry);
            feedEntry = this.read();
          }
        } catch (ex) {
          reject(ex.message);
        }
      })
      .on('end', function () {
        resolve({ feedEntries, feedMeta: this.meta });
      });
  });

  return { feedEntries, feedMeta };
}

async function mapFeedEntriesToModelEntries(options, feedEntries, userRemote) {
  const entries = await Promise.all(
    feedEntries.map(async (feedEntry) => await handleEntry(options, feedEntry, userRemote))
  );
  const filteredEntries = entries.filter((entry) => entry);
  const skippedCount = entries.length - filteredEntries.length;

  return [filteredEntries, skippedCount];
}

async function handleEntry(options, feedEntry, userRemote) {
  const entryId = feedEntry.guid || feedEntry.link || feedEntry.permalink;
  const link = feedEntry.link || feedEntry.permalink;

  const existingModelEntry = await options.getRemoteContent(userRemote.local_username, entryId);

  let dateUpdated = new Date();
  if (feedEntry.date) {
    dateUpdated = new Date(feedEntry.date);
  } else if (feedEntry.pubdate) {
    dateUpdated = new Date(feedEntry.pubdate);
  }

  // We ignore if we already have the item in our DB.
  // Also, we don't keep items that are over options.feedMaxDaysOld.
  if (
    existingModelEntry?.type === 'comment' ||
    (existingModelEntry && +existingModelEntry.updatedAt === +dateUpdated) ||
    dateUpdated < new Date(Date.now() - options.constants.feedMaxDaysOld)
  ) {
    return;
  }

  let view = feedEntry.description || feedEntry.summary;

  const thumbnail = feedEntry['media:group']?.['media:thumbnail']?.['@']['url'];
  if (!view && thumbnail) {
    view = `<a href="${link}" target="_blank" rel="noopener noreferrer"><img src="${thumbnail}" alt="thumbnail" /></a>`;
  }

  view = sanitizeHTML(view);

  // XXX(mime): A shortcoming of feedparser currently is that it doesn't resolve relative urls for feeds that have
  // urls in the content, e.g. kottke.org Fix this hackily for now. It should really be looking at xml:base in the XML.
  const HTML_ATTRIBUTES_WITH_LINKS = [
    'action',
    'background',
    'cite',
    'classid',
    'codebase',
    'href',
    'longdesc',
    'profile',
    'src',
    'usemap',
  ];
  const RELATIVE_REGEXP = new RegExp(`(${HTML_ATTRIBUTES_WITH_LINKS.join('|')})(=['"])/`, 'gi');
  view = view.replace(RELATIVE_REGEXP, `$1$2${userRemote.profile_url}/`);

  // Comments and threads
  let comments_count = 0;
  let comments_updated;
  const atomLinks = feedEntry['atom:link'] ? [feedEntry['atom:link']].flat(1) : [];
  const replies = atomLinks.find((el) => el['@'].rel === 'replies');
  if (replies) {
    comments_count = parseInt(replies['@'].count);
    comments_updated = new Date(replies['@'].updated);
  }
  const thread = feedEntry['thr:in-reply-to']?.['@'].ref;

  // Avatar
  const pocoPhotos = feedEntry['atom:author']?.['poco:photos'];
  const avatar = pocoPhotos && pocoPhotos['poco:value']['#'];

  return {
    id: existingModelEntry?.id || undefined,
    avatar,
    comments_count,
    comments_updated,
    content: '',
    createdAt: feedEntry.pubdate || new Date(),
    creator: feedEntry.author,
    from_user: userRemote.profile_url,
    from_user_remote_id: userRemote.id,
    link,
    post_id: entryId,
    thread,
    title: feedEntry.title || 'untitled',
    to_username: userRemote.local_username,
    type: 'post',
    updatedAt: dateUpdated,
    username: userRemote.username,
    view,
  };
}

class TextStream extends Readable {
  constructor(options, text) {
    super(options);
    this.text = text;
  }

  _read() {
    this.push(this.text);
    this.push(null);
  }
}
