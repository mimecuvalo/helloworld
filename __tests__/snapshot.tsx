import Page404 from 'pages/404';
import { render } from 'util/testing';

it('renders homepage unchanged', () => {
  const { container } = render(<Page404 />);
  expect(container).toMatchSnapshot();
});
