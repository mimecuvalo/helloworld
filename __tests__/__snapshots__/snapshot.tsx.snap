// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders homepage unchanged 1`] = `
<div>
  <div
    class="css-m4uc1f"
  >
    <span
      aria-label="upside down face"
      class="css-184wpat"
      role="img"
    >
      ðŸ™ƒ
    </span>
    <h1
      class="MuiTypography-root MuiTypography-h1 css-1vq31ir-MuiTypography-root"
    >
      <span
        class="notranslate"
      >
        404:
      </span>
       
      <span
        class="i18n-msg"
      >
        not found
      </span>
    </h1>
    <div>
      <span
        class="i18n-msg"
      >
        i'm sorry, dave. i'm afraid i can't do that.
      </span>
      <br />
      <span
        class="i18n-msg"
      >
        try going back to the 
        <a
          class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineHover css-1h36a4r-MuiTypography-root-MuiLink-root"
          href="/"
        >
          beginning
        </a>
        .
      </span>
    </div>
  </div>
</div>
`;
